### PR TITLE
🚨🚨 Migrate `source-zendesk-talk` to low code

### PR DESCRIPTION
**Pull request by ambirdsall**
## What
Converts the `source-zendesk-talk` source to a manifest-based implementation for the low-code CDK. While migrating, I found a logic error with the record parsing of the `ivr_routes` stream, which actually returned records based on the parent `menus` objects. Closes https://github.com/airbytehq/airbyte-internal-issues/issues/6395

## How
1) Defined global properties and at least partial implementations of all streams in the Connector Builder UI
2) Exported the generated YAML to a local file
3) Fixed up some basic inconsistencies and minor errors existing at the time I decided I needed to stop using the Builder: nullable fields in the schemas, removing a partial implementation of `ivr_routes` and `ivr_menus` record parsing via the `transformations` key (which would have been technically possible, but extremely tedious to write and maintain), etc
4) Added a new source class, `SourceZendeskTalkTwo`, alongside the existing python definition along with script definitions, a test catalog for each individual stream, and a testing shell script to run parallel reads of any stream for each. This let me easily compare the results of `spec` and `read` calls (by examining record counts and by pasting output into https://jsondiff.com/ to compare the structure of specs and records) to find and fix inconsistencies
5) Added custom components defining the custom record parsing for `ivr_menus` and `ivr_routes`; while doing so I realized that the existing implementation of `ivr_routes` was incorrect (see below for more details)
6) Added a custom authentication component. I chose not to use `SelectiveAuthenticator` to dispatch to the different possible authenticators because I did not see a way to implement both of the existing ways credentials may be fetched from the config for `BasicHttpAuthenticator` (this might be possible to implement with `SelectiveAuthenticator`, but reading the existing code and documentation I did not see how), and for development speed (there are zero existing usages available for reference).
7) Replaced the unit tests for the old python connector implementation with unit tests for the new custom components
8) Deleted the python definition and its supporting files, renamed the low-code implementation `SourceZendeskTalk`, and deleted the testing scripts and catalogs.

## A bit more on the bug with `ivr_routes`
`ivrs`, `ivr_menus`, and `ivr_routes` all use the same endpoint and response data, but the latter two parse out child records from arrays within that response: each `ivr` contains a list of `menus`, and each `menu` contains a list of `routes`. Here is the old record parsing code from `ivr_routes`:
```python
    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
        """Simply parse json and iterates over root object"""
        ivrs = super().parse_response(response=response, **kwargs)
        for ivr in ivrs:
            for menu in ivr["menus"]:
                for route in ivr["menus"]:
                    yield {"ivr_id": ivr["id"], "ivr_menu_id": menu["id"], **route}
```

Notice that `for menu in ivr["menus"]:` and `for route in ivr["menus"]:` both iterate over the parent `menus` field; the source never iterated through the list of routes within each menu, meaning its records contained extraneous, logically-duplicated data and that its record counts did not accurately reflect the number of route objects returned by the API. I updated the code to actually return one `ivr_routes` record per object defined in some menu's `routes` array.

For the testing account whose credentials are stored in Airbyte's company LastPass, there were 52 `ivrs` with one `menu` each, but only 11 `routes`—most `menus` contained an empty `routes` array. After this fix, the stream returns the expected data given the actual response JSON.

## Schema discrepancies between the old and new versions
- a new `average_transfers_only_time` field was added to the `agents_overview` stream
- a new `transfers_only_time` field was added to `agents_activity`
- a new `capabilities.emergency_address` field was added to `phone_numbers`
- a new `exceeded_queue_wait_time` field added to `calls`
- removed `ivrs.ivr_id`, which does not exist in the API (`ivr_id` only exists as an Airbyte-side addition to the `ivr_menus`/`ivr_routes` "child" streams so users have a reference to the parent object they were extracted from; `ivrs` themselves just use `id`, which was already in the schema)

## 🚨 User Impact 🚨
Breaking changes:
- state changes: the API accepts cursor field filtering _input_ as unix timestamps, but records _output_ datetime strings. The python connector had handrolled code which would always emit record-format-style datetime strings in its state messages, but the low-code CDK normalizes on the query format instead. The python connector errors if given a unix timestamp as its initial state
- removed the `ivr_id` field from the `ivrs` stream schema: it's not part of the record and should not have been in the definition in the first place

**Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*

## Pre-merge Actions
<strong>Updating a connector</strong>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

### CODE REVIEW

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml
Comment (ambirdsall) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:5]: I did of course look at that component, but I chose not to use it because I did not see a way to define both the legacy and updated config structure used with `BasicHttpAuthenticator` the way the python connector did.

**Comment by natikgadzhi** [comment]
Thank you for the detailed writeup of the steps you took to do it. Very interesting! /cc @girarda. 

- It's telling that you chose to implement your own script to compare the output of new and old. That's now how I assumed that should go — perhaps that was easy because regression testing was not ready yet, but this is what regression testing is for.
- It's explicitly OK to break config compatibility — I think we have a thing to implement config migrations. But we can also choose to have custom component if we know there are a lot of connections using this connector today, and the migration will be painful. Let's check the connections count If cloud workspaces are under 10, let's make the breaking change.

Let's run regression tests on this and get this merged quickly! @clnoll another candidate for testing here :D

**Comment by lazebnyi** [comment]
> * It's explicitly OK to break config compatibility — I think we have a thing to implement config migrations. But we can also choose to have custom component if we know there are a lot of connections using this connector today, and the migration will be painful. Let's check the connections count If cloud workspaces are under 10, let's make the breaking change.

From my perspective, I definitely vote for configuration migration + custom component. 

@natikgadzhi Btw, would it be better to calculate the duration of using Airbyte or may be even sync data instead of workspaces qty? Within those 10 workspaces, there could be lightweight connections with newcomers and small amounts of data, as well as heavyweight synchronization processes and users who have been using Airbyte for a long time. Resetting connections or data could be quite painful for them.

**Comment by natikgadzhi** [comment]
@maxi297, @ambirdsall, I get that CAT output can be scary — a session to get this over the finish line, and coach Alex on how to drive a connector migration through completion would be great 👀

**Review by maxi297 — COMMENTED** [review]
Most of the migration seems fine but the devil is in the details!

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml
Comment (ambirdsall) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:833]: You would think, but per [the API docs](https://developer.zendesk.com/api-reference/voice/talk-api/incremental_exports/):
> The incremental export endpoints take a required `start_time` parameter expressed as a Unix epoch time. Example:
>
> `GET /api/v2/channels/voice/stats/incremental/calls?start_time=1547635159`

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/streams.py
Comment (ambirdsall) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/streams.py:147]: This is indeed weird, but that's because the way the Zendesk Talk API handles pagination in their incremental export endpoints is weird (it's designed to support polling). The old code's logic: if the response's `next_page` url uses the same `start_time` as the current page, then there are no more records, which is correct. I changed the implementation to rely on the response's `count` field instead, to be more in line with [the API documentation](https://developer.zendesk.com/api-reference/voice/talk-api/incremental_exports/):
> Unlike other endpoints that return `null` as the value of `next_page` when reaching the last page of the result set, these endpoints return the URL that should be used for the next export. You should use the count property to determine whether more items are available immediately.

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml
Comment (ambirdsall) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:5]: > if we need to revert, we can't. With such a big change as changing the implementation almost completely, it seems rather risky

This was my thinking as well when I opted to dump the existing nonstandard authentication logic into a custom component instead of just keeping the two non-legacy auth options, which _would_ be possible to do in the manifest with `SelectiveAuthenticator`—though given the sparse documentation and lack of existing usages in the repo to use as reference, it might also be a bit slower to build it that way.

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml
Comment (ambirdsall) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:838]: 🤔  The datetime string is the format returned by the API for each record, while the epoch seconds is the format required for the `start_time` param in requests. I didn't specifically test anything around emitted state message format, though, any tips on how to do that @maxi297?

**Comment by maxi297** [comment]
> @maxi297, @ambirdsall, I get that CAT output can be scary — a session to get this over the finish line, and coach Alex on how to drive a connector migration through completion would be great 👀

There is a log mentioning `Request to https://d3v-airbyte.zendesk.com/api/v2/channels/voice/addresses failed with status code 401 and error message None` using `config.json`. Locally, I don't have this issue but it seems like CATs were fixed recently [here](https://connectors.airbyte.com/files/generated_reports/test_summary/source-zendesk-talk/index.html). We can see the error is different but maybe @askarpets fixed this one as well. I would suggest updating the branch and trying again. If it does not work, we can investigate further

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:876]: I don't see the link between the issue mentioned and the API. Right now, if the user sets the `start_date` from the config to `2020-10-15T00:00:00Z`, `2020-10-15T00:00:00Z` will be used a a cursor field and it'll look in the field `2020-10-15T00:00:00Z` for a cursor value

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:881]: This would make sense. I'm just a bit worried because our state could have a cursor value of "1709741421" or "2021-01-01T00:00:00Z" depending on where the value comes from. I would test read with an incremental stream with both those values within the state juste to make sure or check if there is a unit test in DatetimeBasedCursor that validates that both inbound and outbound format can be used for the state

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:143]: I don't know about zendesk but I would do as little change as possible not having this context/those practical reasons

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml
Comment (ambirdsall) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:884]: > check if there is a unit test in DatetimeBasedCursor that validates that both inbound and outbound format can be used for the state

Ah, I was just working with those exact unit tests as part of https://github.com/airbytehq/airbyte/pull/35843: there are unit tests verifying `DatetimeBasedCursor` can handle varying formats, and the time strings get parsed into `datetime` objects for comparison, so changing formats shouldn't affect the logical ordering.

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/components.py
Comment (ambirdsall) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/components.py:1]: Yes, added in fd8311b56085ea158c169b72f9914037b476de07

**Comment by ambirdsall** [comment]
Notes based on prior feedback:
- pagination and incremental field correctness issues fixed
- CATs passing
- schemas compared with a fine-toothed comb; all discrepancies are either fixed or are themselves fixes for errors in the prior definition.

**Comment by maxi297** [comment]
> adding fields

That's all good

> ivrs.greeting_id was moved to ivrs.menus.greeting_id

I'm a bit confused as it seems like `greeting_id` was already under `ivrs.menus` (see [the deleted schema](https://github.com/airbytehq/airbyte/pull/35780/files#diff-9ae4002ff7617af4588ba1957b00e086cbbcb0889a548d17873699e4ba7ed351L16-L18)). Do you mean it was moved to `ivrs.greeting_id` (that is my best assumption based on [the fact that the field seems on the root of ivrs now](https://github.com/airbytehq/airbyte/pull/35780/files#diff-fd45002605750664c19374ea0ddf2a2425ba8dfb505ef3d3351e9f01a86c3b6aR1426-R1429)). This seems to be a breaking change. But assuming the field was never populated, I'm fine with that.

> removed ivrs.ivr_id

It seems like it was never populated anyway so I'm fine with that

**Comment by ambirdsall** [comment]
> it seems like greeting_id was already under ivrs.menus (see [the deleted schema](https://github.com/airbytehq/airbyte/pull/35780/files#diff-9ae4002ff7617af4588ba1957b00e086cbbcb0889a548d17873699e4ba7ed351L16-L18)). Do you mean it was moved to ivrs.greeting_id

I am honestly not sure how I came to make that mistake; it should indeed be (and now is) inside `ivr.menus`.

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:957]: It still seems like this would be a breaking change. The previous version of the connector needs to support the new format if we want to be able to revert.

Comment (ambirdsall) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:957]: Do you consider this a blocker? I.e. do you need to see a PR updating the legacy source definition to handle unix epoch seconds as well as datetime strings merged to `master` before this can be merged?

Comment (ambirdsall) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:957]: Actually, looking more closely at this, I do not think that this would be particularly likely to break anyone's syncs in practice: the updated definition does not define an `end_datetime` for incremental sync, so the record's cursor value—the `updated_at` datetime string, as before—would be used for the state if even a single record were emitted in the sync. I'm not sure that taking the time to avoid this edge case for a connector that only 12 cloud workspaces use would really be striking the best balance between careful work and development speed.

**Comment by natikgadzhi** [comment]
Seeing as this is a widespread issue that we groomed today to address in
the CDK, I think it's fine?

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:957]: I think the question is: can we run a sync on version `0.1.13` with a state the the low-code source generated? If not, this is a breaking change. I'm fine with just documenting the breaking change if it is hard to avoid

File: airbyte-integrations/connectors/source-zendesk-talk/metadata.yaml
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-zendesk-talk/metadata.yaml:31]: @ambirdsall if this change is breaking, you'd have to add the breaking change fields here.
```suggestion
  supportLevel: certified
  releases:
    breakingChanges:
      0.2.0:
        upgradeDeadline: 2024-04-01
        message: "Incremental stream state format has changed in this version."
```

/cc @katmarkham and @ChristoGrab — do we include the detailed breaking change message straight in yaml? I thought it had additional key, right?

**Review by natikgadzhi — COMMENTED** [review]
@ambirdsall, this seems to be sitting for more than a week idle — and it slows us all down. 

Added a suggestion on breaking change messaging. Let's add the breaking change message and ship this out.

File: airbyte-integrations/connectors/source-zendesk-talk/metadata.yaml
Comment (ChristoGrab) [reply airbyte-integrations/connectors/source-zendesk-talk/metadata.yaml:31]: The suggested change is accurate, all thats needed here is the deadline and short blurb. We also need to create a "zendesk-talk-migrations.md" file which will contain the more detailed breakdown of the changes and the migration steps, I've updated the migration template with the info on setting this up: https://docs.google.com/document/d/1SR0S7Xhv2HavV3lFdyzTdk_vWnXJX7XJwv5OkWExOvc/edit

If anything is unclear let me know!

File: airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml
Comment (ambirdsall) [reply airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/manifest.yaml:957]: I manually attempted such a sync, and it triggered a python exception from trying to parse an epoch timestamp with pendulum's datestring code; the breaking change is now documented.

**Comment by ambirdsall** [comment]
> this seems to be sitting for more than a week idle

@natikgadzhi I hope I don't come off too defensive here—I agree wholeheartedly with your prioritization of velocity—but I pushed a fix and replies to the corresponding review comments 3 days ago and have since had to take a sick day. I haven't been satisfied with the time to close this out either, but it has not just been idleness.

In any case, those breaking change updates are now in place, which I hope clears the way for getting the PR approved shortly and ready to ship along with the other breaking migrations.

**Review by maxi297 — APPROVED** [review]
I'm good with this change, thanks for addressing the comments! @bazarnov I would like you to check again so we can have an approval from API Sources as well

**Review by lazebnyi — APPROVED** [review]

File: docs/integrations/sources/zendesk-talk.md
Comment (girarda) [inline docs/integrations/sources/zendesk-talk.md:77]:
```suggestion
| 0.3.0   | 2024-05-06 | [35780](https://github.com/airbytehq/airbyte/pull/35780) | Migrate implementation to low-code CDK                                      |
```

File: airbyte-integrations/connectors/source-zendesk-talk/metadata.yaml
Comment (girarda) [inline airbyte-integrations/connectors/source-zendesk-talk/metadata.yaml:36]:
```suggestion
        upgradeDeadline: "2024-05-31"
```

File: docs/integrations/sources/zendesk-talk-migrations.md
Comment (girarda) [inline docs/integrations/sources/zendesk-talk-migrations.md:3]:
```suggestion
## Upgrading to 1.0.0
```

File: airbyte-integrations/connectors/source-zendesk-talk/metadata.yaml
Comment (girarda) [inline airbyte-integrations/connectors/source-zendesk-talk/metadata.yaml:37]: @katmarkham can you review the breaking change banner when you get a chance? The connector will be good to go once that's done

**Review by katmarkham — APPROVED** [review]

**Pull request merged**
